modules/core/src/test/java/org/apache/ignite/internal/encryption/CacheGroupKeyChangeTest.java

```diff
@@ -0,0 +1,988 @@
+
+        node1.encryption().changeCacheGroupKey(Collections.singleton(cacheName())).get();
+
+        Set<Integer> keys1 = new TreeSet<>(node1.context().encryption().groupKeyIds(grpId));
+        Set<Integer> keys2 = new TreeSet<>(node2.context().encryption().groupKeyIds(grpId));
+
+        assertEquals(2, keys1.size());
+
+        assertEquals(keys1, keys2);
+
+        info("New key was set on all nodes [grpId=" + grpId + ", keys=" + keys1 + "]");
+
+        checkGroupKey(grpId, INITIAL_KEY_ID + 1, MAX_AWAIT_MILLIS);
+
+        stopAllGrids();
+
+        node1 = startGrid(GRID_0);
+        node2 = startGrid(GRID_1);
+
+        node1.cluster().state(ClusterState.ACTIVE);
+
+        // Previous leys must be deleted when the corresponding WAL segment is deleted.
+        try (IgniteDataStreamer<Integer, String> streamer = node1.dataStreamer(cacheName())) {
+            for (int i = node1.cache(cacheName()).size(); i < 500_000; i++) {
+                streamer.addData(i, String.valueOf(i));
+
+                if (i % 1_000 == 0 &&
+                    node1.context().encryption().groupKeyIds(grpId).size() == 1 &&
+                    node2.context().encryption().groupKeyIds(grpId).size() == 1)
+                    break;
+            }
+        }
+
+        assertEquals(1, node1.context().encryption().groupKeyIds(grpId).size());
+        assertEquals(1, node2.context().encryption().groupKeyIds(grpId).size());
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testBasicChangeWithConstantLoad() throws Exception {
+        startTestGrids(true);
+
+        IgniteEx node0 = grid(GRID_0);
+        IgniteEx node1 = grid(GRID_1);
+
+        createEncryptedCache(node0, node1, cacheName(), null);
+
+        forceCheckpoint();
+
+        IgniteInternalCache<Object, Object> cache = node0.cachex(cacheName());
+
+        AtomicInteger cntr = new AtomicInteger(cache.size());
+
+        CountDownLatch startLatch = new CountDownLatch(1);
+
+        final Ignite somenode = node0;
+
+        IgniteInternalFuture<?> loadFut = GridTestUtils.runAsync(() -> {
+            try (IgniteDataStreamer<Integer, String> streamer = somenode.dataStreamer(cacheName())) {
+                while (!Thread.currentThread().isInterrupted()) {
+                    int n = cntr.getAndIncrement();
+
+                    streamer.addData(n, String.valueOf(n));
+
+                    if (n == 5000)
+                        startLatch.countDown();
+                }
+            }
+        });
+
+        startLatch.await(MAX_AWAIT_MILLIS, TimeUnit.MILLISECONDS);
+
+        int grpId = cache.context().groupId();
+
+        node0.encryption().changeCacheGroupKey(Collections.singleton(cacheName())).get(MAX_AWAIT_MILLIS);
+
+        Set<Integer> keys1 = new TreeSet<>(node0.context().encryption().groupKeyIds(grpId));
+        Set<Integer> keys2 = new TreeSet<>(node1.context().encryption().groupKeyIds(grpId));
+
+        assertEquals(keys1, keys2);
```

Comment:
Load future is still working and one of the nodes can already switch WAL segment, do checkpoint and delete key id in background, so test can become flaky.